Provide nix-shell and brief readme instructions

aherrmann opened the pull request:
The nix-expressions for `llvm-hs-pure`, and `llvm-hs` are generated using `cabal2nix`.
The shell refers to a recent commit of the nixpkgs `release-17.03` branch, which contains llvm 4.
Nixpkgs 17.03 should be released soon.

aherrmann committed:
Provide nix-shell and brief readme instructions

sdiehl commented:
Brilliant work. @acowley knows the most about this so should be good to go when he's happy. :+1:

cocreature commented:
Nice! I’m not a nix user so I’ll leave it up to @sdiehl and @acowley to decide when to merge this. One thing that I’d like to know is what kind of maintenance this requires, i.e., when do we need to touch these files? A quick look suggests that the things that only the following things need to be changed:
* The version number of `llvm-hs-pure`.
* The version number of `LLVM` itself
* The `nixpkgs` commit
Is there anything else that needs to be kept up2date?

aherrmann commented:
Thanks for the feedback.

@cocreature
Yes, to update the nixpkgs revision the `rev` and `sha256` attributes would need to be changed in `shell.nix`.
To update LLVM `llvm_4` would need to be changed to e.g. `llvm_5` in `shell.nix`.
If the `llvm-hs` or `llvm-hs-pure` cabal files are changed `cabal2nix` can be used to update the corresponding nix-expression by executing:

```bash
$ (cd llvm-hs-pure && cabal2nix . > default.nix)
$ (cd llvm-hs && cabal2nix -fshared-llvm . > default.nix)
```

acowley commented:
This is really neat, it's very modern! One issue is that the pinning of `nixpkgs` as-is means that there is no binary cache of llvm-4 for macOS users. Perhaps we could add a way to easily disable the pinning to run against the `nixpkgs` in the user's `NIX_PATH` for those who must use the unstable releases?

If I use `<nixpkgs>` in the `shell.nix`, one thing that seems a little funny to me is that it builds `llvm-hs-pure` when entering the shell, but not `llvm-hs` (since we are ultimately using `llvm-hs.env` in `shell.nix`). It's not entirely clear to me if this shell is for working on these packages or using them. To that point, the `cabal` executable is not available in this shell (except via impurity). If I add `cabal-install` as a dependency, then I can build `llvm-hs`. I get a linker failure if I try to test unless I remember to set the `-fshared-llvm` flag, so that should probably go in the instructions since it is in the `default.nix` after all.

After all that, I get a segfault from an OrcJIT test, but I don't think that's related to this change.

Summary:
- Let's make it straightforward to un-pin nixpkgs
- Add a note about `-fshared-llvm` to the instructions
- If this is intended as a development shell, we should add `cabal-install`

aherrmann committed:
nix-shell: Add cabal, don't prebuild llvm-hs-pure

aherrmann committed:
nix-shell: Allow user specified nixpkgs tree

Some users may need to use a different nixpkgs revision
than the one that `shell.nix` specifies.

aherrmann commented:
@acowley Thank you!

> One issue is that the pinning of nixpkgs as-is means that there is no binary cache of llvm-4 for macOS users. Perhaps we could add a way to easily disable the pinning to run against the nixpkgs in the user's NIX_PATH for those who must use the unstable releases?

Absolutely. I have added an argument to the Nix shell. It defaults to the previously pinned version. However, a user can specify a different nixpkgs tree as follows:

```bash
$ nix-shell --arg nixpkgs '<nixpkgs>'
```


> If I use `<nixpkgs>` in the shell.nix, one thing that seems a little funny to me is that it builds llvm-hs-pure when entering the shell, but not llvm-hs (since we are ultimately using llvm-hs.env in shell.nix). It's not entirely clear to me if this shell is for working on these packages or using them.

That's a very good point.
I was following the recommendations of the [nixpkgs manual][1]. However, I agree that it would be better to not pre-build `llvm-hs-pure` to open the Nix shell. I have changed the shell accordingly.

The shell now builds an [ad-hoc environment][2] with all the required dependencies.
The list of dependencies is extracted from the `llvm-hs` and `llvm-hs-pure` derivations, such that `cabal2nix` can still be used to automatically update the Nix shell.
I've also used that opportunity to add cabal to the Nix shell.

[1]: http://nixos.org/nixpkgs/manual/#how-to-build-projects-that-depend-on-each-other
[2]: http://nixos.org/nixpkgs/manual/#how-to-create-ad-hoc-environments-for-nix-shell


> I get a linker failure if I try to test unless I remember to set the
> -fshared-llvm flag, so that should probably go in the instructions since it
> is in the default.nix after all.

It seems that `-fshared-llvm` is now set by default. Do you still get that linker failure?

acowley commented:
This is superb, thank you!

Signed off. +1. 👍. 🙌.

sdiehl commented:
+1 All good. :+1:
